File: rb101/RB101-RB109: Small Problems/Easy 5/letter_swap.rb
# letter_swap.rb
# Given a string of words separated by spaces, write a method that takes this
# string of words and returns a string in which the first and last letters of
# every word are swapped.

# You may assume that every word contains at least one letter, and that the
# string will always contain at least one word. You may also assume that each
# string contains nothing but words and spaces.

# Pseudo-code:

# Data Structure:
# input: string of words separated by spaces
# ouput: string with the first and last letter of each word swapped
# rules:

# Algorithm:
# 1) split the string into separate words and store them in a `words` array
# using the `split(" ")` method.
# 2) iterate over the array, transforming each word and collecting to a new
# `swapped_letters` array using the `each` method.
# 3) rejoin the array into a string using the `join(" ")` method.

def swap(string)
  words = string.split(' ')

  swapped_letters = words.each do |word|
    first = word[0]
    last = word[-1]
    word[0] = last
    word[-1] = first
  end

  swapped_letters.join(' ')
end

puts swap('Oh what a wonderful day it is') == 'hO thaw a londerfuw yad ti si'
puts swap('Abcde') == 'ebcdA'
puts swap('a') == 'a'

puts '-----'

# Launch School solution:

def swap_first_last_characters(word)
  word[0], word[-1] = word[-1], word [0]
  word
end

def swap(words)
  result = words.split.map do |word|
    swap_first_last_characters(word)
  end
  result.join(' ')
end

puts swap('Oh what a wonderful day it is') == 'hO thaw a londerfuw yad ti si'
puts swap('Abcde') == 'ebcdA'
puts swap('a') == 'a'

File: js210/js210_small_problems/easy_3/letter_swap.js
"use strict"

// function swap(text) {
//   let words = text.split(' ');

//   for (let index = 0; index < words.length; index += 1) {
//     let word = words[index];

//     if (word.length > 1) {
//       let firstLetter = word[0];
//       let lastLetter = word[word.length - 1];
//       let wordMiddle = word.slice(1, word.length - 1)
//       words[index] = lastLetter + wordMiddle + firstLetter;
//     }
//   }

//   return words.join(' ');
// }

// console.log(swap('Oh what a wonderful day it is'));
// console.log(swap('Abcde'));
// console.log(swap('a'));

// Further Exploration

function swapFirstLastCharacters(word) {
  if (word.length === 1) {
    return word;
  }

  return word[word.length - 1] + word.slice(1, -1) + word[0];
}

function swap(text) {
  let words = text.split(' ');

  let swapped = words.map(word => swapFirstLastCharacters(word));

  return swapped.join(' ');
}

console.log(swap('Oh what a wonderful day it is'));
console.log(swap('Abcde'));
console.log(swap('a'));
